### ISSUE TITLE
urllib3 should use URLFetch to create HTTPS connections on AppEngine

### ISSUE THREAD

**Issue by micolous** [urllib3/urllib3#583]
AppEngine is listed as a supported platform for `urllib3`.

However, it appears to use the [Sockets API](https://cloud.google.com/appengine/docs/python/sockets/) in order to make HTTPS connections, as a result of it using [`socket` from the standard library](https://docs.python.org/2/library/socket.html) to make the connection.

As a result, this will exhaust quota much faster than if it wrapped the [URLFetch service](https://cloud.google.com/appengine/docs/python/urlfetch/), and is only available to billed (non-free) applications.

Attempting to use it on a free application results in an error.

Additionally, the Sockets API is beta service, and it is [impossible to connect to Google's HTTPS services using the Sockets API](https://cloud.google.com/appengine/docs/python/sockets/#limitations-and-restrictions).

Instead, `urllib3` should use URLFetch by default in order to send HTTPS requests on AppEngine, and [use a similar fallback mechanism to `httplib` in order to fall back to using `socket`](https://cloud.google.com/appengine/docs/python/sockets/#making_httplib_use_sockets).
##

**Comment by shazow** [comment]
Hm it used to work same way httplib does (as we use httplib under the hood). Not quite sure why it stopped working that way. :/ Need to investigate.

**Comment by micolous** [comment]
By the way, this is my request handler for testing:

```python
import webapp2, urllib3      


class IndexHandler(webapp2.RequestHandler):
  def get(self):
    http = urllib3.PoolManager()
    self.response.headers['Content-Type'] = 'text/plain'
    self.response.write('Time to request. ')
    r = http.request('GET', 'https://www.google.com/')
    self.response.write(str(r.status) + '\n')
    self.response.write(r.data)
    return


app = webapp2.WSGIApplication([
  ('/req/', IndexHandler),
], debug=True)
```

If you need anything from the GAE side of things, please ping me and I can assist.

**Comment by shazow** [comment]
@micolous I'd love any help you can offer.

To clarify, normal HTTP connections still work normally, right?

Could you try setting the default HTTPSConnection class to be the UnverifiedHTTPSConnection class? Something like this:

```python
from urllib3.connection import UnverifiedHTTPSConnection
from urllib3 import HTTPSConnectionPool

HTTPSConnectionPool.ConnectionCls = UnverifiedHTTPSConnection

# At this point we can just do what we'd do normally... Such as:
pool = HTTPSConnectionPool("www.google.com")
r = pool.urlopen('GET', '/')
...
```

**Comment by shazow** [comment]
Do we need to [add `validate_certificates` param to urlfetch](https://cloud.google.com/appengine/docs/python/urlfetch/#Python_Secure_connections_and_HTTPS) somehow? (Not sure how we'd do this.)

Also, if `UnverifiedHTTPSConnection` requests on AppEngine are actually verified/secure, we need to get rid of the warning for that scenario. Please confirm this too. :)

**Comment by micolous** [comment]
Normal HTTP connections are fine, and they go out via URLFetch.

I'm still in the progress of testing the code above, at present it's timing out at `pool.urlopen` using the development environment, and I'm having a different issue with the production environment at the moment which I'm attempting to troubleshoot.

You'll need to add the `validate_certificates` parameter to your `fetch` call in order to enable / disable validation explicitly.  It currently defaults to disabled, but will likely change in the future.

**Comment by micolous** [comment]
Another limitation with the URLFetch API will be that you can't explicitly define which certificates you want to use (ie: pin certain certificates, allow different CAs).  You're beholden to AppEngine's trust store.

**Comment by micolous** [comment]
I've now been able to test in a production AppEngine environment with billing.

I had to make one change to the code above, that is, explicitly set the `port` parameter to `HTTPSConnectionPool`.

After this, I can see the Sockets API connection count incrementing (after a short delay).

I ran some tests connecting to HTTPS sites at `facebook.com` and `yahoo.com`, these worked successfully (albeit using the `socket` API).  However, connecting to `google.com` and `googleapis.com` fails with an error.  This is consistent with the `socket` API policies which block access to most Google services.

Error is:

```
HTTPSConnectionPool(host='www.google.com', port=443): Max retries exceeded with url: / (Caused by ProtocolError('Connection aborted.', error(13, 'Permission denied')))
Traceback (most recent call last):
  File "/base/data/home/runtimes/python27/python27_lib/versions/third_party/webapp2-2.3/webapp2.py", line 1511, in __call__
    rv = self.handle_exception(request, response, e)
  File "/base/data/home/runtimes/python27/python27_lib/versions/third_party/webapp2-2.3/webapp2.py", line 1505, in __call__
    rv = self.router.dispatch(request, response)
  File "/base/data/home/runtimes/python27/python27_lib/versions/third_party/webapp2-2.3/webapp2.py", line 1253, in default_dispatcher
    return route.handler_adapter(request, response)
  File "/base/data/home/runtimes/python27/python27_lib/versions/third_party/webapp2-2.3/webapp2.py", line 1077, in __call__
    return handler.dispatch()
  File "/base/data/home/runtimes/python27/python27_lib/versions/third_party/webapp2-2.3/webapp2.py", line 547, in dispatch
    return self.handle_exception(e, self.app.debug)
  File "/base/data/home/runtimes/python27/python27_lib/versions/third_party/webapp2-2.3/webapp2.py", line 545, in dispatch
    return method(*args, **kwargs)
  File "/base/data/home/apps/s~MYAPPID/2.XXXXXXXXX/req.py", line 16, in get
    r = pool.urlopen('GET', '/')
  File "/base/data/home/apps/s~MYAPPID/2.XXXXXXXXX/urllib3/connectionpool.py", line 617, in urlopen
    release_conn=release_conn, **response_kw)
  File "/base/data/home/apps/s~MYAPPID/2.XXXXXXXXX/urllib3/connectionpool.py", line 617, in urlopen
    release_conn=release_conn, **response_kw)
  File "/base/data/home/apps/s~MYAPPID/2.XXXXXXXXX/urllib3/connectionpool.py", line 617, in urlopen
    release_conn=release_conn, **response_kw)
  File "/base/data/home/apps/s~MYAPPID/2.XXXXXXXXX/urllib3/connectionpool.py", line 597, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/base/data/home/apps/s~MYAPPID/2.XXXXXXXXX/urllib3/util/retry.py", line 271, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
MaxRetryError: HTTPSConnectionPool(host='www.google.com', port=443): Max retries exceeded with url: / (Caused by ProtocolError('Connection aborted.', error(13, 'Permission denied')))
```

**Comment by Lukasa** [comment]
So i think this problem happened when we took over the connection logic so that we could apply socket options. This meant we replaced `httplib`'s `connect` function with one of our own, that explicitly uses sockets.

The best we could do, I think, is to *conditionally* throw back to `httplib` if we're running on GAE and socket options aren't  being explicitly set.

Note, however, that GAE breaks a lot of our other assumptions too: timeouts will *at best* not work the same and at worst not work at all, for example.

**Comment by shazow** [comment]
:/

**Comment by sigmavirus24** [comment]
Didn't we have someone working on making all things GAE better? Was it @webmaven?

**Comment by shazow** [comment]
Another idea for resolving this: We could have an AppEngineConnection base class to fall back onto iff we detect we're in the AppEngine environment. It would be super-simple and just use urlfetch.

**Comment by webmaven** [comment]
@sigmavirus24:
> Didn't we have someone working on making all things GAE better? Was it @webmaven?

Yes, but I'm not part of the Google Cloud Platform team (or even employed by Google), and ran out of free time to push that effort forward.

Good advocates who *are* on that team are @proppy and @jonparrott.

**Comment by jonparrott** [comment]
I'm not familiar with the internal of URLFetch or urllib3, but I'm willing to help. @shazow recommended having an AppEngineConnection class. Is that the easiest way to isolate all appengine specific code? I would highly recommend this as the way forward. I can provide reliable code for detecting the app engine environment.

I also see that there aren't any appengine specific tests. Is that something that's desired? I can look into getting those together.

There is also the new, non-sandboxed managed vms runtime which should, by all means, be a standard environment as it is docker on compute engine underneath. However, it seems you guys wish to support application on the standard runtime using the free tier.

**Comment by shazow** [comment]
> I'm not familiar with the internal of URLFetch or urllib3, but I'm willing to help. @shazow recommended having an AppEngineConnection class. Is that the easiest way to isolate all appengine specific code? I would highly recommend this as the way forward. I can provide reliable code for detecting the app engine environment.

I think that's the best way to start with. Once that is ready, we may need to move things around a bit to make it easier to specify which kind of connection to use for each pool.

> I also see that there aren't any appengine specific tests. Is that something that's desired? I can look into getting those together.

+1, would be ideal if we had somekind of automated testing environment, not sure if it's feasible to integrate it into travisci with the rest of our suite.

> There is also the new, non-sandboxed managed vms runtime which should, by all means, be a standard environment as it is docker on compute engine underneath. However, it seems you guys wish to support application on the standard runtime using the free tier.

Correct, not worried about managed VMs. All of our AppEngine-related complaints come from people who use AppEngine without sockets.

**Comment by jonparrott** [comment]
> +1, would be ideal if we had somekind of automated testing environment, not sure if it's feasible to integrate it into travisci with the rest of our suite.

We did this in our repository for [python documentation samples](https://github.com/GoogleCloudPlatform/python-docs-samples), so it should be possible. Where would be the best place to put these?

**Comment by shazow** [comment]
Maybe a `test/appengine/...`. Feel free to sketch out some ideas in a PR and we can iterate from there. :)

**Comment by jonparrott** [comment]
Sounds good, I will try to get some basic tests in this weekend. Keep in mind that I may be delayed by our internal review process.

**Comment by shazow** [comment]
@jonparrott Thanks a bunch, much appreciated!

**Comment by shazow** [comment]
FYI, @jonparrott started work on AppEngine tests in #631. Reviews/suggestions appreciated. :)

**Comment by jonparrott** [comment]
I'm tentatively taking this one on by creating a AppEngineConnectionPool class. If someone wants to do it and I haven't updated, feel free to steal it from me.

**Comment by shazow** [comment]
@jonparrott Any news on this?

**Comment by jonparrott** [comment]
@shazow I'll be trying to prioritize this next week, thanks for your patience!

**Comment by shazow** [comment]
:custard:

**Comment by jonparrott** [comment]
Early implementation of the GAE pool class is in PR #664.

**Comment by shazow** [comment]
We just merged a contrib module that does this (should be in the upcoming v1.11 release). We'd appreciate some testing of it before we migrate it as a native feature in a release or two. :)

Docs here: https://urllib3.readthedocs.org/en/latest/contrib.html#google-app-engine

**Comment by webmaven** [comment]
Awesome, @jonparrott, thanks!

**Comment by sigmavirus24** [comment]
Should this be closed now that we have the AppEngineManager?

**Comment by shazow** [comment]
I was thinking of holding it open until we merge it into the mainline urllib3 flow, which I hope to be sooner than later. @jonparrott Will you have some time in the coming weeks to work on the next steps for this?

**Comment by jonparrott** [comment]
Potentially in August. What steps are left for this to happen on my end? The manager is currently the highest-level construct that I'm aware of, how do you expect the automatic handoff to work from a user perspective?

**Comment by shazow** [comment]
Yea, the abstraction above Manager is still being talked about, and hopefully we'll have something put together in August to work with.

Until then, we could do something like SNI support in Py2: https://urllib3.readthedocs.org/en/latest/contrib.html#contrib-modules
Add a `urllib3.contrib.appengine.inject_into_poolmanager()` which replaces/monkeypatches PoolManager iff it's in the AppEngine environment.

Not sure if it's the best idea.

Also we could move it from contrib into `urllib3/appengine/` or somesuch, make sure it has full test coverage and documentation.

**Comment by jonparrott** [comment]
I'd highly prefer *not* monkeypatching, just a personal preference.

What's needed to move from contrib to main? Test coverage and docs?

**Comment by shazow** [comment]
> I'd highly prefer not monkeypatching, just a personal preference.

Then we may need to punt making it automagic until we have another abstraction layer to work with.

> What's needed to move from contrib to main? Test coverage and docs?

Yes. Though I guess it doesn't *have* to move out of contrib. (Sorry I'm kinda waffly on these things right now for no good reason.)

**Comment by jonparrott** [comment]
> Though I guess it doesn't have to move out of contrib

That's fine, but I'd still like it to be up to standards.

**Comment by hwdavidward** [comment]
Any news on this?

**Comment by sigmavirus24** [comment]
@hwdavidward this has been part of the contrib sub-module for quite some time.

**Comment by hwdavidward** [comment]
@sigmavirus24 could you tell me why I get "InsecurePlatformWarning" messages when I use the python datadog api on GAE then? I was under the impression that this issue related to this ticket. 

`/base/data/home/apps/s~/1-8-6.404526562025782184/lib/urllib3/util/ssl_.py:137: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings`

**Comment by sigmavirus24** [comment]
I don't believe they are related. If they were related you wouldn't be able to make any connection. In this case you're being warned that your version of Python doesn't make an SSLContext object available to us. The documentation linked from the warning you pasted explains this quite clearly, I thought.

**Comment by davidparsson** [comment]
If this were to be implemented, I think it still should be possible to use the current approach since AppEngine's URLFetch API does _not_ support all use-cases currently supported. One example is client-side certificates.

**Comment by micolous** [comment]
**Note:** I don't presently work on Google Cloud.

**URLFetch is not required** for the [Python 3.7 Standard Runtime](https://cloud.google.com/appengine/docs/standard/python3/quickstart) (which uses a different sandboxing system) or any of the [Flexible Environment runtimes](https://cloud.google.com/appengine/docs/flexible/python/) (which use Compute Engine).

The weirdness is because under the hood, App Engine (on Python 2.7 Standard Runtime) is [sending your HTTP(S) requests as an RPC](https://landing.google.com/sre/sre-book/chapters/production-environment/#our-software-infrastructure-XQs4iw) to a service that makes HTTP(S) requests or your behalf.  This service is also used for stuff like Google's web crawler, which doesn't care too much about implementing things like TLS client certificates or certificate pinning.

I've seen plenty of App Engine users shoot themselves in the foot because `urllib3` (or `requests`) used Sockets API without them realising and then blow through the (at the time) very small quota, or otherwise try to access Google API services and not figure out why it doesn't work.  The overwhelming majority of the time, URLFetch provided all the features needed (with the _possible_ exception of being harder to use `requests`).

_But, it's now 2019, so you should absolutely be using one of the Python 3.7 runtimes, and [the clock is literally ticking to migrate](https://pythonclock.org/)._  That doesn't represent the status of Python 2.7 support on App Engine, but you can expect more Python libraries to stop working from 2020, as they start ripping out compatibility code. In some cases, that decision may be made by another library that a library you depend on uses (so you may have a surprise).

With my reporter hat on, I'm going to _close_ the original issue as [I'm satisfied that `urllib3` now has a way to handle this](https://urllib3.readthedocs.io/en/latest/advanced-usage.html#google-app-engine) and I don't think there is a lot of value in doing _more_ than this given that Python 3.7 Standard Runtime doesn't need it, and Python 2.7 should be left to sail quietly into the night. :)

Thank you @theacodes and @shazow for getting in a work-around!
